File: frontend/src/lib/services/api/sections.ts
/** Section servisi — endpoint listesi: docs/architecture.md §15 */

import { api } from './client';
import type { SectionDto, SectionTreeResponse } from '$lib/types';

export interface CreateSectionInput {
	name: string;
	code?: string;
	type?: string;
	parent_id?: string | null;
}

export interface BulkSectionInput {
	parent_id?: string | null;
	start_index: number;
	count: number;
	name_format: string;
	code_format?: string;
	type?: string;
}

export interface CloneSectionInput {
	count: number;
	name_format?: string;
	code_format?: string;
}

export const sectionService = {
	tree(projectId: string): Promise<SectionTreeResponse> {
		return api.get(`/projects/${projectId}/sections/tree`);
	},

	create(projectId: string, input: CreateSectionInput): Promise<SectionDto> {
		return api.post(`/projects/${projectId}/sections`, input);
	},

	update(
		sectionId: string,
		patch: { name?: string; code?: string | null; type?: string | null }
	): Promise<SectionDto> {
		return api.patch(`/sections/${sectionId}`, patch);
	},

	/** Soft delete; alt bölümü olanlar 409 alır. */
	remove(sectionId: string): Promise<{ ok: boolean }> {
		return api.del(`/sections/${sectionId}`);
	},

	bulkCreate(projectId: string, input: BulkSectionInput): Promise<SectionDto[]> {
		return api.post(`/projects/${projectId}/sections/bulk`, input);
	},

	/** Alt ağacı `count` kez kopyalar; ad otomatik türetilir ("1. Kat" → "2. Kat"). */
	clone(sectionId: string, input: CloneSectionInput): Promise<SectionDto[]> {
		return api.post(`/sections/${sectionId}/clone`, input);
	}
};

File: backend/src/api/routes/sections.rs
//! Section endpoint'leri (docs/architecture.md §15).

use crate::api::error::ApiResult;
use crate::api::extractor::AuthUser;
use crate::api::state::AppState;
use crate::application::dto::{
    BulkSectionRequest, CloneSectionRequest, CreateSectionRequest, SectionDto,
    SectionTreeResponse, UpdateSectionRequest,
};
use crate::application::services::section_service as service;
use axum::extract::{Path, State};
use axum::Json;
use uuid::Uuid;

/// GET /api/v1/projects/{project_id}/sections/tree
pub async fn tree(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(project_id): Path<Uuid>,
) -> ApiResult<Json<SectionTreeResponse>> {
    Ok(Json(service::tree(&state.pool, &auth.user, project_id).await?))
}

/// POST /api/v1/projects/{project_id}/sections
pub async fn create(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(project_id): Path<Uuid>,
    Json(req): Json<CreateSectionRequest>,
) -> ApiResult<(axum::http::StatusCode, Json<SectionDto>)> {
    let section = service::create(&state.pool, &auth.user, project_id, req).await?;
    Ok((axum::http::StatusCode::CREATED, Json(section.into())))
}

/// POST /api/v1/projects/{project_id}/sections/bulk
pub async fn bulk_create(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(project_id): Path<Uuid>,
    Json(req): Json<BulkSectionRequest>,
) -> ApiResult<(axum::http::StatusCode, Json<Vec<SectionDto>>)> {
    let created = service::bulk_create(&state.pool, &auth.user, project_id, req).await?;
    Ok((
        axum::http::StatusCode::CREATED,
        Json(created.into_iter().map(Into::into).collect()),
    ))
}

/// POST /api/v1/sections/{section_id}/clone
pub async fn clone(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(section_id): Path<Uuid>,
    Json(req): Json<CloneSectionRequest>,
) -> ApiResult<(axum::http::StatusCode, Json<Vec<SectionDto>>)> {
    let created = service::clone(&state.pool, &auth.user, section_id, req).await?;
    Ok((
        axum::http::StatusCode::CREATED,
        Json(created.into_iter().map(Into::into).collect()),
    ))
}

/// PATCH /api/v1/sections/{section_id}
pub async fn update(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(section_id): Path<Uuid>,
    Json(req): Json<UpdateSectionRequest>,
) -> ApiResult<Json<SectionDto>> {
    Ok(Json(
        service::update(&state.pool, &auth.user, section_id, req)
            .await?
            .into(),
    ))
}

/// DELETE /api/v1/sections/{section_id} — soft delete (confirm'lu UI aksiyonu)
pub async fn delete(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(section_id): Path<Uuid>,
) -> ApiResult<Json<serde_json::Value>> {
    service::delete(&state.pool, &auth.user, section_id).await?;
    Ok(Json(serde_json::json!({ "ok": true })))
}
